### ISSUE TITLE
[Help Wanted] [MM-11327] Restrict Teams by Email

### ISSUE THREAD

**Issue by esethna** [mattermost/mattermost-server#9111]
If you're interested please comment here and come [join our "Contributors" community channel](https://pre-release.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://pre-release.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://docs.mattermost.com/guides/developer.html), specifically for [machine setup](https://docs.mattermost.com/developer/developer-setup.html) and for [developer workflow](https://docs.mattermost.com/developer/developer-flow.html).

----

**Notes**: [Jira ticket](https://mattermost.atlassian.net/browse/MM-11327)

Based on request submitted in [Uservoice|https://mattermost.uservoice.com/forums/306457-general/suggestions/33014368-restrict-team-by-email]

**Comment by gvengel** [comment]
I started putting together [code for a PR](/gvengel/mattermost-server/compare/master...gvengel:restrict-teams-by-email) which leverages the existing team.AllowedDomains model in master. I was unaware of it at the time, but it looks like my code basically mirrors the first commit from PR #6954. 

PR #6954 was closed pending "proper restriction between teams.". I believe that work has actually been completed, or at least it is now possible to restrict DM's to a team. (Settings > Users and Teams > Enable users to open Direct Message channels with).

I'll start looking at porting #6954 into my branch; however, I'm new to the Mattermost code base, Golang, and React, so progress may be slow going. :)
